### PR TITLE
.DEFINEDINSTR

### PR DESCRIPTION
**Pull request by Movax12**
This is very close to the .DEFINEDMACRO function, but for instructions. It does need to check UbiquitousIdents however. I removed the .FEATURE requirement since it was suggested we don't need to for .DEFINEDMACRO.

### CODE REVIEW

File: src/ca65/expr.c [hunk 1]
```diff
@@ -417,6 +418,34 @@ static ExprNode* FuncDefined (void)
 
 
 
+static ExprNode* FuncDefinedInstr (void)
+/* Handle the .DEFINEDINSTR builtin function */
+{
+    int Instr = 0;
```
Comment (greg-king5) [inline src/ca65/expr.c]: `FindInstruction()` returns `-1` if its argument isn't an instruction.

File: src/ca65/expr.c [hunk 2]
```diff
@@ -417,6 +418,34 @@ static ExprNode* FuncDefined (void)
...
+            /* Macros CAN be instructions, so check for them first */
+            if (FindMacro (&CurTok.SVal) == 0) {
+                Instr = FindInstruction (&CurTok.SVal);
+            }
+        } else {
+            /* Macros and symbols may NOT use the names of instructions, so just check for the instruction */
+            Instr = FindInstruction (&CurTok.SVal);
+        }
+    } else {
+        Error ("Idenitifier expected.");
```
Comment (greg-king5) [inline src/ca65/expr.c]: The usual typo. :wink:

**Comment by greg-king5** [comment]
We don't define instructions; they're built in.  I think that this function should be named something like `.isinst()` (for "is instruction").

I noticed that the description of `.DEF, .DEFINED` was put in the wrong section.  It isn't a control command; it's a function.  It should be moved to the `Pseudo functions` section (and, that section should be renamed just `Functions`) -- but, do it in a separate Pull Request.

I mentioned `.defined()` because `.ISINST` (and, `.DEFINEDMACRO`) also need to be moved.

**Comment by Movax12** [comment]
How about `.isinstr()` ? I'd like the 'r' to be included.

**Comment by greg-king5** [comment]
"instr" makes me think of "in string" (some programming languages have a string function with that name).

File: src/ca65/expr.c (hunk 1 above)
Comment (MonteCarlos) [inline src/ca65/expr.c]: Is

.definedopstr()

a possible choice? For "operation code"

MC

Am 19.05.2015 um 11:49 schrieb greg-king5:

**Comment by greg-king5** [comment]
`.isopstr()` looks like a viable suggestion to me.

**Comment by Movax12** [comment]
I can't help but see string when I see `str` now. I think I like `.ismnem()` and `.ismnemonic().` better, adn I think we should include both.

**Comment by greg-king5** [comment]
Implementing both spellings is a good idea.

**Comment by Movax12** [comment]
Changes to documentation to come.

**Comment by oliverschmidt** [comment]
Hi,

Maybe the comment I just made in https://github.com/cc65/cc65/pull/158 is true here too...

Regards,
Oliver

**Comment by Movax12** [comment]
Thanks for checking on these requests. I am still working on them - I have just moved and have been busy, but I plan to get back to work on this soon. Incidentally, I think this specific request is complete from my perspective.

**Comment by oliverschmidt** [comment]
Hi,

> Thanks for checking on these requests. I am still working on them - I have
> just moved and have been busy, but I plan to get back to work on this soon.
>
> Good to know. I just wanted to make sure that not "everybody" is waiting
for me without me noticing it...

> Incidentally, I think this specific request is complete from my
> perspective.
>
I see. Let's wait if Greg has something to add.

Regards,
Oliver

>

**Comment by greg-king5** [comment]
I agree with "Movax12".  In fact, as far as I can tell, _both_ PR #158 and this one are finished.  Both of them are ready for Oliver's action.

**Comment by oliverschmidt** [comment]
Hi Greg,

> In fact, as far as I can tell, both PR #158 and this one are finished.

Thanks for your fedback :-)

> Both of them are ready for Oliver's action.

I will act accordingly.

Regards,
Oliver

**Pull request merged**